components/eamxx/src/share/field/field_utils.hpp

```diff
@@ -43,6 +43,69 @@ void randomize (const Field& f, Engine& engine, PDF&& pdf)
   impl::randomize<ST>(f,engine,pdf);
 }
 
+// Compute a random perturbation of a field for all view entries whose
+// level index satisfies the mask.
+// Input:
+//   - f:                  Field to perturbed. Required to have level midpoint
+//                         tag as last dimension.
+//   - engine:             Random number engine.
+//   - pdf:                Random number distribution where a random value (say,
+//                         pertval) is taken s.t.
+//                           field_view(i0,...,iN) *= (1 + pertval)
+//   - base_seed:          Seed used for creating the engine input.
+//   - level_mask:         Mask (size of the level dimension of f) where f(i0,...,k) is
+//                         perturbed if level_mask(k)=true
+//   - dof_gids:           Field containing global DoF IDs for columns of f (if applicable)
+template<typename Engine, typename PDF, typename MaskType>
+void perturb (const Field& f,
+              Engine& engine,
+              PDF&& pdf,
+              const int base_seed,
+              const MaskType& level_mask,
+              const Field& dof_gids = Field())
+{
+  EKAT_REQUIRE_MSG(f.is_allocated(),
+       	           "Error! Cannot perturb the values of a field not yet allocated.\n");
+
+  // Deduce scalar type from pdf
+  using ST = decltype(pdf(engine));
+
+  // Check compatibility between PDF and field data type
+  const auto data_type = f.data_type();
+  EKAT_REQUIRE_MSG((std::is_same_v<ST,int> && data_type==DataType::IntType) or
+                   (std::is_same_v<ST,float> && data_type==DataType::FloatType) or
+                   (std::is_same_v<ST,double> && data_type==DataType::DoubleType),
+                   "Error! Field data type incompatible with input PDF.\n");
+
+  using namespace ShortFieldTagsNames;
+  const auto& fl = f.get_header().get_identifier().get_layout();
+
+  // Field we are perturbing should have a level midpoint dimension,
+  // and it is required to be the last dimension
+  EKAT_REQUIRE_MSG(fl.has_tag(LEV),
```

Comment:
You could condensate these two checks in one, and add a printout of the field layout. Something like

```suggestion
  EKAT_REQUIRE_MSG(fl.rank()>0 and (fl.tags().back()==LEV or fl.tags().back()==ILEV),
    "Error! Trying to perturb a field with no vertical levels dimension dimension.\n"
    "  - field name: " + f.name() + "\n"
    "  - field layout: " + to_string(fl) + "\n");
```